src/verify.rs

```diff
@@ -0,0 +1,160 @@
+    raw_metadata: &RawSignedMetadata<D, M>,
+    threshold: u32,
+    authorized_keys: I,
+) -> Result<Verified<M>>
+where
+    D: DataInterchange,
+    M: Metadata,
+    I: IntoIterator<Item = &'a PublicKey>,
+{
+    if threshold < 1 {
+        return Err(Error::VerificationFailure(
+            "Threshold must be strictly greater than zero".into(),
+        ));
+    }
+
+    let authorized_keys = authorized_keys
+        .into_iter()
+        .map(|k| (k.key_id(), k))
+        .collect::<HashMap<&KeyId, &PublicKey>>();
+
+    // Extract the signatures and canonicalize the bytes.
+    let (signatures, canonical_bytes) = {
+        #[derive(Deserialize)]
+        pub struct SignedMetadata<D: DataInterchange> {
+            signatures: Vec<Signature>,
+            signed: D::RawData,
+        }
+
+        let unverified: SignedMetadata<D> = D::from_slice(raw_metadata.as_bytes())?;
+
+        if unverified.signatures.is_empty() {
+            return Err(Error::VerificationFailure(
+                "The metadata was not signed with any authorized keys.".into(),
+            ));
+        }
+
+        let canonical_bytes = D::canonicalize(&unverified.signed)?;
+        (unverified.signatures, canonical_bytes)
+    };
+
+    let mut signatures_needed = threshold;
+
+    // Create a key_id->signature map to deduplicate the key_ids.
+    let signatures = signatures
+        .iter()
+        .map(|sig| (sig.key_id(), sig))
+        .collect::<HashMap<&KeyId, &Signature>>();
+
+    for (key_id, sig) in signatures {
+        match authorized_keys.get(key_id) {
+            Some(ref pub_key) => match pub_key.verify(&canonical_bytes, sig) {
+                Ok(()) => {
+                    debug!("Good signature from key ID {:?}", pub_key.key_id());
+                    signatures_needed -= 1;
+                }
+                Err(e) => {
+                    warn!("Bad signature from key ID {:?}: {:?}", pub_key.key_id(), e);
+                }
+            },
+            None => {
+                warn!(
+                    "Key ID {:?} was not found in the set of authorized keys.",
+                    sig.key_id()
+                );
+            }
+        }
+        if signatures_needed == 0 {
+            break;
+        }
+    }
+
+    if signatures_needed > 0 {
+        return Err(Error::VerificationFailure(format!(
+            "Signature threshold not met: {}/{}",
+            threshold - signatures_needed,
+            threshold
+        )));
+    }
+
+    // Everything looks good so deserialize the metadata.
+    //
+    // Note: Canonicalization (or any other transformation of data) could modify or filter out
+    // information about the data. Therefore, while we've confirmed the canonical bytes are signed,
+    // we shouldn't interpret this as if the raw bytes were signed. So we deserialize from the
+    // `canonical_bytes`, rather than from `raw_meta.as_bytes()`.
+    let verified_metadata = D::from_slice(&canonical_bytes)?;
```

Comment:
As we talked about offline, it is more expensive, but what we're trying to avoid is assuming just because we've verified the signatures of the canonical bytes, we can't assume that the `SignedMetadata` is signed. It's possible that canonicalization could change the metadata in some critical manner, such as removing a field. It's possible that the presence of that field later on could compromise the security of rust-tuf. So it's best if we throw away the raw metadata we used to extract out the signatures.

Reply:
Thanks for clarifying, I hadn't considered that D::RawData may strip out data when parsing generically parsing metadata.  serde_json::Value may not, but DataInterchange may be implemented outside this crate, so we shouldn't be assuming it works as intended.